### PR TITLE
CP-51988: Fix functions not work for remote_pool repo

### PR DESCRIPTION
**Pull request by BengangY**
1. `remote_pool` repo doesn't support periodic sync updates.
2. Periodic sync updates should be auto-disabled when calling `set_repositories` and `add_repository` for `remote_pool` repo.
4. Update UT.

### CODE REVIEW

File: ocaml/xapi/xapi_pool.ml [hunk 1]
```diff
@@ -3453,7 +3459,10 @@ let set_repositories ~__context ~self ~value =
   Db.Pool.set_repositories ~__context ~self ~value ;
   if Db.Pool.get_repositories ~__context ~self = [] then
     Db.Pool.set_last_update_sync ~__context ~self ~value:Date.epoch ;
-  disable_auto_update_sync_for_bundle_repo ~__context ~self ~repos:value
+  if not (can_periodic_sync_updates ~__context ~repos:value) then (
+    Pool_periodic_update_sync.set_enabled ~__context ~value:false ;
+    Db.Pool.set_update_sync_enabled ~__context ~self ~value:false
+  )
```
Comment (gangj) [inline ocaml/xapi/xapi_pool.ml:3465]: Would be good to have a function to avoid duplicates in `add_repository`.

**Review by gangj — APPROVED** [review]

File: ocaml/idl/datamodel_errors.ml [hunk 2]
```diff
@@ -1905,7 +1905,13 @@ let _ =
   error Api_errors.bundle_repo_not_enabled []
     ~doc:"Cannot sync bundle as the bundle repository is not enabled." () ;
   error Api_errors.can_not_sync_updates []
-    ~doc:"Cannot sync updates as the bundle repository is enabled." () ;
+    ~doc:"The current enabled repositories are not supported to sync updates."
```
Comment (lindig) [inline ocaml/idl/datamodel_errors.ml:1908]: Do you mean: the currently enabled repositories do not support synchronization of updates?

File: ocaml/idl/datamodel_errors.ml [hunk 3]
```diff
@@ -1905,7 +1905,13 @@ let _ =
...
+    () ;
+  error Api_errors.can_not_periodic_sync_updates []
+    ~doc:
+      "The current enabled repositories are not supported to periodic sync \
```
Comment (lindig) [inline ocaml/idl/datamodel_errors.ml:1912]: The currently enabled repositories do not support periodic automatic updates.

File: ocaml/idl/datamodel_errors.ml (hunk 2 above)
Comment (BengangY) [reply ocaml/idl/datamodel_errors.ml]: Yes. I updated it.

File: ocaml/xapi/xapi_pool.ml [hunk 4]
```diff
@@ -3406,21 +3406,33 @@ let enable_tls_verification ~__context =
   | Some self ->
       Xapi_cluster_host.set_tls_config ~__context ~self ~verify:true
 
-let contains_bundle_repo ~__context ~repos =
-  List.exists
-    (fun repo -> Db.Repository.get_origin ~__context ~self:repo = `bundle)
-    repos
-
 let assert_single_bundle_repo_can_be_enabled ~__context ~repos =
```
Comment (gangj) [inline ocaml/xapi/xapi_pool.ml:3409]: I think we also need to have the same kind of restriction for `remote_pool` repository, or how about change it to: `assert_single_update_repo_can_be_enabled`.

File: ocaml/idl/datamodel_errors.ml [hunk 5]
```diff
@@ -1905,11 +1905,19 @@ let _ =
...
-       repository of the pool."
+      "The currently enabled repositories do not support synchronization of \
+       updates."
+    () ;
+  error Api_errors.can_not_periodic_sync_updates []
+    ~doc:
+      "The currently enabled repositories do not support periodic automatic \
+       updates."
+    () ;
+  error Api_errors.repo_should_be_single_enabled []
```
Comment (gangj) [inline ocaml/idl/datamodel_errors.ml:1917]: Add the type of the repo, `remote` or `remote_pool` as the error parameter.

**Comment by xueqingz** [comment]
Java SDK build failed on:
```
    /**
     * If the bundle repository or remote_pool repository is enabled, it should be the only one enabled repository of the pool.
     */
    public static class RepoShouldBeSingleEnabled extends XenAPIException {
        public final String repo-type;

        /**
         * Create a new RepoShouldBeSingleEnabled
         */
        public RepoShouldBeSingleEnabled(String repo-type) {
            super("If the bundle repository or remote_pool repository is enabled, it should be the only one enabled repository of the pool.");
            this.repo-type = repo-type;
        }
    }
``` 

Consider change the name "repo-type" to "repo_type" ?

File: ocaml/idl/datamodel_errors.ml [hunk 6]
```diff
@@ -1905,11 +1905,19 @@ let _ =
...
-       repository of the pool."
+      "The currently enabled repositories do not support synchronization of \
+       updates."
+    () ;
+  error Api_errors.can_not_periodic_sync_updates []
+    ~doc:
+      "The currently enabled repositories do not support periodic automatic \
+       updates."
+    () ;
+  error Api_errors.repo_should_be_single_enabled ["repo-type"]
```
Comment (gangj) [inline ocaml/idl/datamodel_errors.ml:1917]: repo_type

**Comment by BengangY** [comment]
> Java SDK build failed on:
> 
> ```
>     /**
>      * If the bundle repository or remote_pool repository is enabled, it should be the only one enabled repository of the pool.
>      */
>     public static class RepoShouldBeSingleEnabled extends XenAPIException {
>         public final String repo-type;
> 
>         /**
>          * Create a new RepoShouldBeSingleEnabled
>          */
>         public RepoShouldBeSingleEnabled(String repo-type) {
>             super("If the bundle repository or remote_pool repository is enabled, it should be the only one enabled repository of the pool.");
>             this.repo-type = repo-type;
>         }
>     }
> ```
> 
> Consider change the name "repo-type" to "repo_type" ?

Updated to "repo_type" and it passed now.

File: ocaml/xapi-consts/api_errors.ml [hunk 7]
```diff
@@ -1322,8 +1322,9 @@ let bundle_repo_not_enabled = add_error "BUNDLE_REPO_NOT_ENABLED"
 
 let can_not_sync_updates = add_error "CAN_NOT_SYNC_UPDATES"
 
-let bundle_repo_should_be_single_enabled =
-  add_error "BUNDLE_REPO_SHOULD_BE_SINGLE_ENABLED"
+let can_not_periodic_sync_updates = add_error "CAN_NOT_PERIODIC_SYNC_UPDATES"
+
+let repo_should_be_single_enabled = add_error "REPO_SHOULD_BE_SINGLE_ENABLED"
```
Comment (lindig) [inline ocaml/xapi-consts/api_errors.ml:1327]: The wording is a bit odd. Does this mean that only one repo should be enabled? "repo should be single enabled" is not grammatical and ambiguous. "repo should be the single one enabled" could be something I would guess.

**Review by lindig — APPROVED** [review]

File: ocaml/xapi-consts/api_errors.ml (hunk 7 above)
Comment (BengangY) [reply ocaml/xapi-consts/api_errors.ml:1327]: > Does this mean that only one repo should be enabled?
If the bundle repository or remote_pool repository is enabled, it should be the only one enabled repository of the pool. Or else, multiple repos (only remote) can be enabled.

File: ocaml/tests/test_pool_repository.ml [hunk 8]
```diff
@@ -14,101 +14,159 @@
...
-      Xapi_pool.add_repository ~__context ~self ~value:ref_remote
-    )
+      Xapi_pool.set_repositories ~__context ~self ~value:[ref_remote_pool] ;
+      Alcotest.check_raises "test_set_repositories_1"
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_enabled
+            , [
+                Record_util.repo_origin_to_string `bundle
+              ; Record_util.repo_origin_to_string `remote_pool
```
Comment (gangj) [inline ocaml/tests/test_pool_repository.ml:49]: No `remote_pool` type of repo is set here

File: ocaml/tests/test_pool_repository.ml [hunk 9]
```diff
@@ -14,101 +14,159 @@
...
+        (fun () ->
+          Xapi_pool.set_repositories ~__context ~self
+            ~value:[ref_remote_pool; ref_bundle]
+        ) ;
+      Alcotest.check_raises "test_set_repositories_3"
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_enabled
+            , [
+                Record_util.repo_origin_to_string `bundle
```
Comment (gangj) [inline ocaml/tests/test_pool_repository.ml:76]: No `bundle` type of repo is set here.

File: ocaml/tests/test_pool_repository.ml [hunk 10]
```diff
@@ -14,101 +14,159 @@
...
-    )
+let test_add_repository () =
+  on_repositories (fun __context self ref_remote ref_bundle ref_remote_pool ->
+      Alcotest.check_raises "test_add_repository_1"
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_enabled
+            , [
+                Record_util.repo_origin_to_string `bundle
+              ; Record_util.repo_origin_to_string `remote_pool
```
Comment (gangj) [inline ocaml/tests/test_pool_repository.ml:95]: No `remote_pool` type of repo is set.

File: ocaml/tests/test_pool_repository.ml [hunk 11]
```diff
@@ -14,101 +14,159 @@
...
+        (fun () ->
+          Xapi_pool.set_repositories ~__context ~self ~value:[ref_remote] ;
+          Xapi_pool.add_repository ~__context ~self ~value:ref_bundle
+        ) ;
+      Alcotest.check_raises "test_add_repository_2"
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_enabled
+            , [
+                Record_util.repo_origin_to_string `bundle
```
Comment (gangj) [inline ocaml/tests/test_pool_repository.ml:108]: No `bundle` type of repo is set

File: ocaml/xapi-consts/api_errors.ml (hunk 7 above)
Comment (gangj) [reply ocaml/xapi-consts/api_errors.ml:1327]: I agree renaming it to `repo_should_be_the_single_one_enabled` would be more clearer.

File: ocaml/xapi-cli-server/record_util.ml [hunk 12]
```diff
@@ -551,3 +551,11 @@ let vm_placement_policy_of_string a =
       `anti_affinity
   | s ->
       record_failure "Invalid VM placement policy, got %s" s
+
+let repo_origin_to_string = function
```
Comment (robhoes) [inline ocaml/xapi-cli-server/record_util.ml:555]: I don't think we need this and we can rely on the auto-generated function now. See the comment on top of this file.

File: ocaml/tests/test_pool_repository.ml [hunk 13]
```diff
@@ -14,101 +14,156 @@
...
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_one_enabled
+            , [
+                String.concat ", "
+                  [
+                    Record_util.origin_to_string `bundle
+                  ; Record_util.origin_to_string `remote_pool
+                  ]
+              ]
```
Comment (gangj) [inline ocaml/tests/test_pool_repository.ml:64]: [`bundel; `remote_pool] |> List.map Record_util.origin_to_string
And the same for others.

File: ocaml/xapi/xapi_pool.ml [hunk 14]
```diff
@@ -3406,21 +3406,54 @@ let enable_tls_verification ~__context =
...
+               Some origin
+           | `remote ->
+               None
+       )
+    |> List.fold_left
+         (fun acc origin -> if List.mem origin acc then acc else origin :: acc)
+         []
+  in
+  match (repos, origins) with
+  | _ :: _ :: _, _ :: _ ->
```
Comment (gangj) [inline ocaml/xapi/xapi_pool.ml:3424]: only
`match origins with | _::_`
is enough?

File: ocaml/xapi/xapi_pool.ml [hunk 15]
```diff
@@ -3406,21 +3406,54 @@ let enable_tls_verification ~__context =
...
+      raise
+        Api_errors.(
+          Server_error
+            ( repo_should_be_single_one_enabled
+            , [
+                origins
+                |> List.map Record_util.origin_to_string
+                |> String.concat ", "
+              ]
```
[the remark this answers is not in the document]
Comment (BengangY) [reply ocaml/xapi/xapi_pool.ml:3433]: Need to concat the list to a string.

File: ocaml/xapi/xapi_pool.ml (hunk 14 above)
Comment (BengangY) [reply ocaml/xapi/xapi_pool.ml:3424]: Yes, it's enough. We just check the count of origins is larger than 0.

File: ocaml/tests/test_pool_repository.ml (hunk 13 above)
Comment (gangj) [reply ocaml/tests/test_pool_repository.ml]: I see several errors are using this format, like `Api_errors.pvs_site_contains_running_proxies`, `Api_errors.network_contains_vif` and `Api_errors.network_contains_pif`, would you please help to confirm if this is indeed a problem we should not follow? Thank you!

File: ocaml/idl/datamodel_errors.ml [hunk 16]
```diff
@@ -1905,11 +1905,19 @@ let _ =
...
-       repository of the pool."
+      "The currently enabled repositories do not support synchronization of \
+       updates."
+    () ;
+  error Api_errors.can_not_periodic_sync_updates []
+    ~doc:
+      "The currently enabled repositories do not support periodic automatic \
+       updates."
+    () ;
+  error Api_errors.repo_should_be_single_one_enabled ["repo_type"]
```
Comment (gangj) [inline ocaml/idl/datamodel_errors.ml:1917]: repo_types

File: ocaml/tests/test_pool_repository.ml (hunk 13 above)
Comment (BengangY) [reply ocaml/tests/test_pool_repository.ml]: I followed the existing errors and updated it.

**Review by gangj — APPROVED** [review]

**Pull request merged**